File: examples/dekita/.claude/hooks/issue-reference-check.py
#!/usr/bin/env python3
"""存在しないIssue参照をブロックする。

Why:
    存在しないIssue番号を参照すると、後から追跡できなくなり
    誤った情報がコメントに残る。参照前に存在確認を強制する。

What:
    - gh pr comment/gh issue comment等のコメント投稿を検出
    - コメント本文から#1234形式のIssue参照を抽出
    - gh issue viewで存在確認、不存在ならブロック

Remarks:
    - ブロック型フック（存在しないIssue参照時はブロック）
    - PreToolUse:Bashで発火（gh pr/issue comment、gh api replies）
    - Closes/Fixes/Resolvesパターンは除外（Issue作成用途）
    - cross-repoコマンド（--repo）はスキップ（他リポジトリ検証困難）

Changelog:
    - silenvx/dekita#2059: フック追加
"""

import json
import re
import subprocess
import sys

from lib.constants import TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.results import make_approve_result, make_block_result
from lib.session import parse_hook_input

# Pattern to detect Issue references: #1234
ISSUE_REF_PATTERN = re.compile(r"#(\d+)")

# Pattern to detect Closes/Fixes/Resolves keywords (allowed even for non-existent Issues)
CLOSES_PATTERN = re.compile(r"(?:closes?|fixes?|resolves?)\s*#\d+", re.IGNORECASE)

# Pattern to detect PR references (allowed, validated differently from Issues)
PR_PATTERN = re.compile(r"PR\s*#\d+", re.IGNORECASE)

# Patterns to detect comment commands
COMMENT_COMMAND_PATTERNS = [
    # gh pr comment
    r"gh\s+pr\s+comment\b",
    # gh api .../replies
    r"gh\s+api\s+.*?/replies\b",
    # gh api graphql with addPullRequestReviewThreadReply
    r"gh\s+api\s+graphql.*addPullRequestReviewThreadReply",
    # gh issue comment
    r"gh\s+issue\s+comment\b",
]


def is_comment_command(command: str) -> bool:
    """Check if the command is a comment-posting command."""
    return any(
        re.search(pattern, command, re.IGNORECASE | re.DOTALL)
        for pattern in COMMENT_COMMAND_PATTERNS
    )


def read_body_from_file(file_path: str) -> str | None:
    """Read comment body from a file."""
    try:
        with open(file_path, encoding="utf-8") as f:
            return f.read()
    except (FileNotFoundError, PermissionError, OSError):
        return None


def extract_comment_body(command: str) -> str | None:
    """Extract comment body from command.

    Handles:
    - HEREDOC patterns (checked first, highest priority)
    - --body-file <file> (reads from file)
    - -F body=@<file> (reads from file, gh api format)
    - -b "message" or --body "message"
    - -b $'message' (bash quoting)
    - GraphQL body parameter
    """
    # HEREDOC pattern (check first - highest priority)
    match = re.search(r"<<['\"]?EOF['\"]?\s*\n?(.*?)EOF", command, re.DOTALL)
    if match:
        return match.group(1)

    # --body-file <file> (gh pr comment / gh issue comment)
    match = re.search(r"--body-file\s+[\"']?([^\s\"']+)[\"']?", command)
    if match:
        return read_body_from_file(match.group(1))

    # -F body=@<file> (gh api format, file reference)
    match = re.search(r"-F\s+body=@[\"']?([^\s\"']+)[\"']?", command)
    if match:
        return read_body_from_file(match.group(1))

    # -f body="message" or -F body="message" (gh api inline format)
    match = re.search(r"-[fF]\s+body=[\"'](.+?)[\"']", command, re.DOTALL)
    if match:
        return match.group(1)

    # -b/--body $'message' (bash $'' quoting)
    match = re.search(r"(?:-b|--body)\s+\$'(.+?)'", command, re.DOTALL)
    if match:
        return match.group(1)

    # -b/--body "message" or -b/--body 'message'
    # Exclude $(cat <<) patterns to avoid matching HEREDOC wrappers
    match = re.search(r"(?:-b|--body)\s+[\"']([^$].+?)[\"']", command, re.DOTALL)
    if match:
        return match.group(1)

    # Fallback for simple quoted strings starting with $
    match = re.search(r"(?:-b|--body)\s+[\"'](.+?)[\"']", command, re.DOTALL)
    if match and "<<" not in match.group(1):
        return match.group(1)

    # GraphQL body parameter (in mutation)
    match = re.search(r'body:\s*["\'](.+?)["\']', command, re.DOTALL)
    if match:
        return match.group(1)

    return None


def check_issue_exists(issue_number: int) -> bool:
    """Check if an Issue exists using gh CLI.

    Returns True (fail-open) for any error except explicit "not found".
    """
    try:
        result = subprocess.run(
            ["gh", "issue", "view", str(issue_number), "--json", "number"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
        )
        if result.returncode == 0:
            return True

        # Only return False if stderr indicates "not found" (Issue doesn't exist)
        # Other errors (network, auth, API) should fail-open
        stderr_lower = result.stderr.lower()
        if "not found" in stderr_lower or "could not resolve" in stderr_lower:
            return False

        # Unknown error - fail-open to avoid blocking valid comments
        return True
    except Exception:
        # On any error, assume Issue exists to avoid false positives (fail-open)
        return True


def extract_issue_references(text: str) -> list[int]:
    """Extract Issue numbers from text, excluding Closes/Fixes and PR patterns."""
    # Remove Closes/Fixes patterns from text first
    text_without_closes = CLOSES_PATTERN.sub("", text)

    # Remove PR #xxx patterns (PRs are validated differently from Issues)
    text_without_pr = PR_PATTERN.sub("", text_without_closes)

    # Find all Issue references
    matches = ISSUE_REF_PATTERN.findall(text_without_pr)
    return [int(m) for m in matches]


def main():
    """Entry point for the Issue reference check hook."""
    try:
        input_json = parse_hook_input()
        tool_name = input_json.get("tool_name", "")

        # Only check Bash commands
        if tool_name != "Bash":
            result = make_approve_result("issue-reference-check")
            log_hook_execution("issue-reference-check", "approve", "not Bash")
            print(json.dumps(result))
            return

        tool_input = input_json.get("tool_input") or {}
        command = tool_input.get("command") or ""

        # Check if it's a comment command
        if not is_comment_command(command):
            result = make_approve_result("issue-reference-check")
            log_hook_execution("issue-reference-check", "approve", "not comment command")
            print(json.dumps(result))
            return

        # Skip validation for cross-repo commands (--repo flag)
        # We can't reliably check Issues in other repos, so fail-open
        if re.search(r"--repo\s+\S+", command):
            result = make_approve_result("issue-reference-check")
            log_hook_execution("issue-reference-check", "approve", "cross-repo command")
            print(json.dumps(result))
            return

        # Extract comment body
        body = extract_comment_body(command)
        if not body:
            result = make_approve_result("issue-reference-check")
            log_hook_execution("issue-reference-check", "approve", "no body found")
            print(json.dumps(result))
            return

        # Extract Issue references
        issue_numbers = extract_issue_references(body)
        if not issue_numbers:
            result = make_approve_result("issue-reference-check")
            log_hook_execution("issue-reference-check", "approve", "no Issue refs")
            print(json.dumps(result))
            return

        # Check each Issue exists
        non_existent = []
        for issue_num in issue_numbers:
            if not check_issue_exists(issue_num):
                non_existent.append(issue_num)

        if non_existent:
            issues_str = ", ".join(f"#{n}" for n in non_existent)
            reason = (
                f"存在しないIssueを参照しています: {issues_str}\n\n"
                "Issueを参照する前に、まず `gh issue create` で作成してください。\n"
                "Issue作成後、実際のIssue番号を使用してコメントを再投稿してください。\n\n"
                "背景: Issue番号を推測して参照すると、\n"
                "実際のIssue番号との不一致が発生します（Issue #2059）。"
            )
            result = make_block_result("issue-reference-check", reason)
            log_hook_execution(
                "issue-reference-check",
                "block",
                reason,
                {"non_existent": non_existent, "command": command[:100]},
            )
            print(json.dumps(result))
            return

        # All Issues exist
        result = make_approve_result("issue-reference-check")
        log_hook_execution(
            "issue-reference-check",
            "approve",
            f"verified: {issue_numbers}",
        )
        print(json.dumps(result))

    except Exception as e:
        # On error, approve to avoid blocking legitimate commands
        print(f"[issue-reference-check] Hook error: {e}", file=sys.stderr)
        result = make_approve_result("issue-reference-check", f"Hook error: {e}")
        log_hook_execution("issue-reference-check", "approve", f"Hook error: {e}")
        print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/issue_reference_check.ts
#!/usr/bin/env bun
/**
 * 存在しないIssue参照をブロックする。
 *
 * Why:
 *   存在しないIssue番号を参照すると、後から追跡できなくなり
 *   誤った情報がコメントに残る。参照前に存在確認を強制する。
 *
 * What:
 *   - gh pr comment/gh issue comment等のコメント投稿を検出
 *   - コメント本文から#1234形式のIssue参照を抽出
 *   - gh issue viewで存在確認、不存在ならブロック
 *
 * Remarks:
 *   - ブロック型フック（存在しないIssue参照時はブロック）
 *   - PreToolUse:Bashで発火（gh pr/issue comment、gh api replies）
 *   - Closes/Fixes/Resolvesパターンは除外（Issue作成用途）
 *   - cross-repoコマンド（--repo）はスキップ（他リポジトリ検証困難）
 *
 * Changelog:
 *   - silenvx/dekita#2059: フック追加
 *   - silenvx/dekita#2932: TypeScriptに移行
 */

import { extractCommentBody } from "../lib/command";
import { TIMEOUT_MEDIUM } from "../lib/constants";
import { logHookExecution } from "../lib/logging";
import { makeApproveResult, makeBlockResult } from "../lib/results";
import { parseHookInput } from "../lib/session";

const HOOK_NAME = "issue-reference-check";

// Pattern to detect Issue references: #1234
const ISSUE_REF_PATTERN = /#(\d+)/g;

// Pattern to detect Closes/Fixes/Resolves keywords (allowed even for non-existent Issues)
const CLOSES_PATTERN = /(?:closes?|fixes?|resolves?)\s*#\d+/gi;

// Pattern to detect PR references (allowed, validated differently from Issues)
const PR_PATTERN = /PR\s*#\d+/gi;

// Patterns to detect comment commands
const COMMENT_COMMAND_PATTERNS = [
  // gh pr comment
  /gh\s+pr\s+comment\b/i,
  // gh api .../replies
  /gh\s+api\s+.*?\/replies\b/i,
  // gh api graphql with addPullRequestReviewThreadReply
  /gh\s+api\s+graphql.*addPullRequestReviewThreadReply/is,
  // gh issue comment
  /gh\s+issue\s+comment\b/i,
];

/**
 * Check if the command is a comment-posting command.
 */
export function isCommentCommand(command: string): boolean {
  return COMMENT_COMMAND_PATTERNS.some((pattern) => pattern.test(command));
}

/**
 * Check if an Issue exists using gh CLI.
 *
 * Returns True (fail-open) for any error except explicit "not found".
 */
export async function checkIssueExists(issueNumber: number): Promise<boolean> {
  try {
    const proc = Bun.spawn(["gh", "issue", "view", String(issueNumber), "--json", "number"], {
      stdout: "pipe",
      stderr: "pipe",
    });

    const timeoutPromise = new Promise<never>((_, reject) => {
      setTimeout(() => reject(new Error("timeout")), TIMEOUT_MEDIUM * 1000);
    });

    const exitCode = await Promise.race([proc.exited, timeoutPromise]);

    if (exitCode === 0) {
      return true;
    }

    // Only return False if stderr indicates "not found" (Issue doesn't exist)
    // Other errors (network, auth, API) should fail-open
    const stderr = await new Response(proc.stderr).text();
    const stderrLower = stderr.toLowerCase();
    if (stderrLower.includes("not found") || stderrLower.includes("could not resolve")) {
      return false;
    }

    // Unknown error - fail-open to avoid blocking valid comments
    return true;
  } catch {
    // On any error, assume Issue exists to avoid false positives (fail-open)
    return true;
  }
}

/**
 * Extract Issue numbers from text, excluding Closes/Fixes and PR patterns.
 */
export function extractIssueReferences(text: string): number[] {
  // Remove Closes/Fixes patterns from text first
  const textWithoutCloses = text.replace(CLOSES_PATTERN, "");

  // Remove PR #xxx patterns (PRs are validated differently from Issues)
  const textWithoutPr = textWithoutCloses.replace(PR_PATTERN, "");

  // Find all Issue references
  const matches: number[] = [];
  for (const match of textWithoutPr.matchAll(ISSUE_REF_PATTERN)) {
    matches.push(Number.parseInt(match[1], 10));
  }

  return [...new Set(matches)]; // Deduplicate
}

async function main(): Promise<void> {
  let inputData: Awaited<ReturnType<typeof parseHookInput>>;
  let sessionId: string | undefined;
  try {
    inputData = await parseHookInput();
    sessionId = inputData.session_id;
  } catch {
    // Fail open
    return;
  }

  const toolName = inputData.tool_name ?? "";

  // Only check Bash commands
  if (toolName !== "Bash") {
    await logHookExecution(HOOK_NAME, "approve", "not Bash", undefined, { sessionId });
    console.log(JSON.stringify(makeApproveResult(HOOK_NAME)));
    return;
  }

  const toolInput = inputData.tool_input ?? {};
  const command = (toolInput as { command?: string }).command ?? "";

  // Check if it's a comment command
  if (!isCommentCommand(command)) {
    await logHookExecution(HOOK_NAME, "approve", "not comment command", undefined, { sessionId });
    console.log(JSON.stringify(makeApproveResult(HOOK_NAME)));
    return;
  }

  // Skip validation for cross-repo commands (--repo flag)
  // We can't reliably check Issues in other repos, so fail-open
  if (/--repo\s+\S+/.test(command)) {
    await logHookExecution(HOOK_NAME, "approve", "cross-repo command", undefined, { sessionId });
    console.log(JSON.stringify(makeApproveResult(HOOK_NAME)));
    return;
  }

  // Extract comment body
  const body = await extractCommentBody(command);
  if (!body) {
    await logHookExecution(HOOK_NAME, "approve", "no body found", undefined, { sessionId });
    console.log(JSON.stringify(makeApproveResult(HOOK_NAME)));
    return;
  }

  // Extract Issue references
  const issueNumbers = extractIssueReferences(body);
  if (issueNumbers.length === 0) {
    await logHookExecution(HOOK_NAME, "approve", "no Issue refs", undefined, { sessionId });
    console.log(JSON.stringify(makeApproveResult(HOOK_NAME)));
    return;
  }

  // Check each Issue exists
  const nonExistent: number[] = [];
  for (const issueNum of issueNumbers) {
    const exists = await checkIssueExists(issueNum);
    if (!exists) {
      nonExistent.push(issueNum);
    }
  }

  if (nonExistent.length > 0) {
    const issuesStr = nonExistent.map((n) => `#${n}`).join(", ");
    const reason = `存在しないIssueを参照しています: ${issuesStr}

Issueを参照する前に、まず \`gh issue create\` で作成してください。
Issue作成後、実際のIssue番号を使用してコメントを再投稿してください。

背景: Issue番号を推測して参照すると、
実際のIssue番号との不一致が発生します（Issue #2059）。`;

    await logHookExecution(
      HOOK_NAME,
      "block",
      reason,
      {
        non_existent: nonExistent,
        command: command.slice(0, 100),
      },
      { sessionId },
    );
    console.log(JSON.stringify(makeBlockResult(HOOK_NAME, reason)));
    return;
  }

  // All Issues exist
  await logHookExecution(
    HOOK_NAME,
    "approve",
    `verified: ${JSON.stringify(issueNumbers)}`,
    undefined,
    { sessionId },
  );
  console.log(JSON.stringify(makeApproveResult(HOOK_NAME)));
}

if (import.meta.main) {
  main().catch((error) => {
    console.error(`[${HOOK_NAME}] Error:`, error);
    process.exit(0); // Fail open
  });
}
